uefi/src/boot.rs

```diff
@@ -0,0 +1,101 @@
+//! UEFI boot services.
+//!
+//! These functions will panic if called after exiting boot services.
+
+use crate::data_types::PhysicalAddress;
+use core::ptr::{self, NonNull};
+use uefi::{table, Result, StatusExt};
+
+#[cfg(doc)]
+use uefi::Status;
+
+pub use uefi::table::boot::AllocateType;
+pub use uefi_raw::table::boot::MemoryType;
+
+fn boot_services_raw_panicking() -> NonNull<uefi_raw::table::boot::BootServices> {
+    let st = table::system_table_raw_panicking();
+    // SAFETY: valid per requirements of `set_system_table`.
+    let st = unsafe { st.as_ref() };
+    NonNull::new(st.boot_services).expect("boot services are not active")
+}
+
+/// Allocates memory pages from the system.
+///
+/// UEFI OS loaders should allocate memory of the type `LoaderData`. A `u64`
+/// is returned even on 32-bit platforms because some hardware configurations
+/// like Intel PAE enable 64-bit physical addressing on a 32-bit processor.
+///
+/// # Errors
+///
+/// * [`Status::OUT_OF_RESOURCES`]: allocation failed.
+/// * [`Status::INVALID_PARAMETER`]: `mem_ty` is [`MemoryType::PERSISTENT_MEMORY`],
+///   [`MemoryType::UNACCEPTED`], or in the range [`MemoryType::MAX`]`..=0x6fff_ffff`.
+/// * [`Status::NOT_FOUND`]: the requested pages could not be found.
+pub fn allocate_pages(
+    ty: AllocateType,
+    mem_ty: MemoryType,
+    count: usize,
+) -> Result<PhysicalAddress> {
```

Comment:
Seems like an API inconsistency with `allocate_pool`, which returns `Result<NonNull<u8>>`.

Reply:
That's true. This inconsistency is present in our `BootServices` methods and in the UEFI spec; `AllocatePool` produces a `void*` pointer, while `AllocatePages` produces a `EFI_PHYSICAL_ADDRESS`. Notably, `EFI_PHYSICAL_ADDRESS` is always a `UINT64` rather than being pointer sized, so there is a real difference in behavior. The docstring notes this in the context of ia32 PAE (I copied it from the `BootServices` method, the text  was originally added in [this commit](https://github.com/rust-osdev/uefi-rs/pull/65/commits/d8faa60b91ae7b1f91569c53e825b1063ac10f0d)).

That said, I think in almost all cases the first thing a caller is going to do with the `EFI_PHYSICAL_ADDRESS` is cast it to a pointer, and I can't currently think of any real-world counter example. So, while I don't feel very strongly in either direction, I don't mind converting this to return `NonNull<u8>` instead of `PhysicalAddress`.

Reply:
Updated. For full consistency, I also changed both free functions to use `NonNull<u8>` as well.